# Chapter 5 — Data Preparation and Feature Engineering

This notebook turns the raw M5 Forecasting Accuracy data into a tidy panel
DataFrame with calendar, event, SNAP, price, lag, rolling, and cyclical
features. The output is `processed_train_m5.parquet`, consumed by the next
two notebooks.

Run `download_m5.py` first to fetch the CSVs (CC-BY 4.0).

In [1]:
import numpy as np
import pandas as pd

## Load the raw files

In [2]:
calendar = pd.read_csv("calendar.csv", parse_dates=["date"])
sales_wide = pd.read_csv("sales_train_validation.csv")
prices = pd.read_csv("sell_prices.csv")

print("calendar  ", calendar.shape)
print("sales_wide", sales_wide.shape)
print("prices    ", prices.shape)
calendar.head()

calendar   (1969, 14)
sales_wide (30490, 1919)
prices     (6841121, 4)


,date,wm_yr_wk,weekday,wday,month,year,d,event_name_1,event_type_1,event_name_2,event_type_2,snap_CA,snap_TX,snap_WI
0,2011-01-29,11101,Saturday,1,1,2011,d_1,NaN,NaN,NaN,NaN,0,0,0
1,2011-01-30,11101,Sunday,2,1,2011,d_2,NaN,NaN,NaN,NaN,0,0,0
2,2011-01-31,11101,Monday,3,1,2011,d_3,NaN,NaN,NaN,NaN,0,0,0
3,2011-02-01,11101,Tuesday,4,2,2011,d_4,NaN,NaN,NaN,NaN,1,1,0
4,2011-02-02,11101,Wednesday,5,2,2011,d_5,NaN,NaN,NaN,NaN,1,0,1


## Subsample for runnability

The full panel is 30,490 series × 1,913 days ≈ 58 M rows. For teaching, we
sample 100 random items and keep all 10 stores; the resulting ~1,000 series
matches the order of magnitude of the original Rossmann setup and runs
comfortably on a laptop. Bump `SAMPLE_ITEMS` to scale up.

In [3]:
SAMPLE_ITEMS = 100
SAMPLE_SEED = 42

rng = np.random.RandomState(SAMPLE_SEED)
items_all = sales_wide["item_id"].unique()
sample_items = rng.choice(items_all, size=SAMPLE_ITEMS, replace=False)

sales_wide = sales_wide[sales_wide["item_id"].isin(sample_items)].copy()
print(f"After sampling: {sales_wide.shape[0]} series")

After sampling: 1000 series


## Reshape to long format and merge

`sales_train_validation.csv` is wide (one column per day, `d_1`…`d_1913`).
We melt it to long, then merge the calendar metadata (date, events, SNAP
flags) and the weekly sell prices.

In [4]:
id_vars = ["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"]
day_cols = [c for c in sales_wide.columns if c.startswith("d_")]

sales_long = sales_wide.melt(
    id_vars=id_vars,
    value_vars=day_cols,
    var_name="d",
    value_name="sales",
)

df = sales_long.merge(calendar, on="d", how="left")
df = df.rename(columns={"date": "Date"})
df = df.merge(prices, on=["store_id", "item_id", "wm_yr_wk"], how="left")
print(df.shape)
df.head(3)

(1913000, 22)


,id,item_id,dept_id,cat_id,store_id,state_id,d,sales,Date,wm_yr_wk,...,month,year,event_name_1,event_type_1,event_name_2,event_type_2,snap_CA,snap_TX,snap_WI,sell_price
0,HOBBIES_1_045_CA_1_validation,HOBBIES_1_045,HOBBIES_1,HOBBIES,CA_1,CA,d_1,0,2011-01-29,11101,...,1,2011,NaN,NaN,NaN,NaN,0,0,0,NaN
1,HOBBIES_1_052_CA_1_validation,HOBBIES_1_052,HOBBIES_1,HOBBIES,CA_1,CA,d_1,0,2011-01-29,11101,...,1,2011,NaN,NaN,NaN,NaN,0,0,0,NaN
2,HOBBIES_1_053_CA_1_validation,HOBBIES_1_053,HOBBIES_1,HOBBIES,CA_1,CA,d_1,0,2011-01-29,11101,...,1,2011,NaN,NaN,NaN,NaN,0,0,0,NaN


## Drop pre-launch rows

A `NaN` in `sell_price` means the product wasn't on sale at that store yet:
filling those with 0 or a median would invent demand the model would learn
to predict. We drop them.

In [5]:
pre_launch = df["sell_price"].isna().sum()
df = df.dropna(subset=["sell_price"]).reset_index(drop=True)
print(f"Dropped {pre_launch:,} pre-launch rows; remaining shape: {df.shape}")

Dropped 447,874 pre-launch rows; remaining shape: (1465126, 22)


## Calendar features

Some of these columns already exist in `calendar.csv` (year, month, wday)
but we recompute from `Date` so the column names stay consistent with the
rest of the chapter.

In [6]:
df["Year"] = df["Date"].dt.year
df["Month"] = df["Date"].dt.month
df["Day"] = df["Date"].dt.day
df["DayOfWeek"] = df["Date"].dt.dayofweek           # Monday = 0, Sunday = 6
df["WeekOfYear"] = df["Date"].dt.isocalendar().week.astype(int)
df["DayOfYear"] = df["Date"].dt.dayofyear

## Event and SNAP features

`event_name_1` lists named events on each day (SuperBowl, Easter, Christmas,
…). The three `snap_*` columns flag SNAP food-benefit redemption days per
state, so we resolve the right column based on each row's `state_id`.

In [7]:
df["IsEvent"] = df["event_name_1"].notna().astype(int)

df["snap_active"] = np.select(
    [df["state_id"] == "CA", df["state_id"] == "TX", df["state_id"] == "WI"],
    [df["snap_CA"], df["snap_TX"], df["snap_WI"]],
    default=0,
).astype(int)

## Price-derived promotion

M5 doesn't have a Promo column, but `sell_price` does the work. A markdown
shows up as today's price falling below 95% of the 30-day rolling median.
The `shift(1)` keeps today out of the median window so the feature is
computable at prediction time.

In [8]:
df = df.sort_values(["id", "Date"]).reset_index(drop=True)

df["price_30d_median"] = (
    df.groupby("id")["sell_price"]
      .transform(lambda x: x.shift(1).rolling(window=30, min_periods=7).median())
)
df["Promo"] = (
    (df["sell_price"] < 0.95 * df["price_30d_median"])
    & df["price_30d_median"].notna()
).astype(int)

print("Promo days:", df["Promo"].sum(), "of", len(df))

Promo days: 7174 of 1465126


## Lag features

Sales from yesterday, last week, two weeks ago, and four weeks ago. We
group by `id` (item-store pair) before shifting so values don't bleed
across series.

In [9]:
LAGS = [1, 7, 14, 28]
for lag in LAGS:
    df[f"sales_lag_{lag}"] = df.groupby("id")["sales"].shift(lag)

## Rolling window statistics

Rolling means and standard deviations of past sales. The `shift(1)` before
`.rolling()` keeps the current day out of the window — without it the
feature would peek at today's target.

In [10]:
ROLLING_WINDOWS = [7, 14, 28]
for w in ROLLING_WINDOWS:
    df[f"sales_rolling_mean_{w}"] = (
        df.groupby("id")["sales"]
          .transform(lambda x: x.shift(1).rolling(window=w).mean())
    )
    df[f"sales_rolling_std_{w}"] = (
        df.groupby("id")["sales"]
          .transform(lambda x: x.shift(1).rolling(window=w).std())
    )

## Cyclical encoding

Sin/cos pairs preserve cyclical adjacency (December next to January, Sunday
next to Monday) — important for any model that doesn't natively know about
the wrap-around.

In [11]:
def encode_cyclical(df, col, max_val):
    df[f"{col}_sin"] = np.sin(2 * np.pi * df[col] / max_val)
    df[f"{col}_cos"] = np.cos(2 * np.pi * df[col] / max_val)
    return df

df = encode_cyclical(df, "DayOfWeek", 7)
df = encode_cyclical(df, "Month", 12)
# 'Day' is day of month: 31 is a fixed approximation. For month-aware
# encoding, look up the month length with calendar.monthrange(year, month)[1].
df = encode_cyclical(df, "Day", 31)
df = encode_cyclical(df, "WeekOfYear", 52)

## Interaction features

In [12]:
df["IsWeekend"] = df["DayOfWeek"].isin([5, 6]).astype(int)
df["PromoOnWeekend"] = df["Promo"] * df["IsWeekend"]
df["EventOnWeekend"] = df["IsEvent"] * df["IsWeekend"]

## Drop warm-up rows and save

Rows where the longest rolling window has no history still hold `NaN` lag
and rolling values. We drop them — never `bfill`, which would copy future
sales backwards into the feature columns.

In [13]:
longest_roll = f"sales_rolling_mean_{max(ROLLING_WINDOWS)}"
pre = len(df)
df = df.dropna(subset=[longest_roll]).reset_index(drop=True)
print(f"Dropped {pre - len(df):,} warm-up rows; remaining: {df.shape}")

Dropped 28,000 warm-up rows; remaining: (1437126, 53)


In [14]:
drop_cols = [
    "price_30d_median",
    "event_name_1", "event_name_2", "event_type_1", "event_type_2",
    "snap_CA", "snap_TX", "snap_WI",
    "weekday", "wday", "month", "year", "wm_yr_wk", "d",
]
df = df.drop(columns=[c for c in drop_cols if c in df.columns])

print("Final shape:", df.shape)
print("Dtype counts:")
print(df.dtypes.value_counts())

df.to_parquet("processed_train_m5.parquet", index=False)
print("Saved processed_train_m5.parquet")

Final shape: (1437126, 39)
Dtype counts:
float64           19
int64              8
object             6
int32              5
datetime64[ns]     1
Name: count, dtype: int64


Saved processed_train_m5.parquet
